lib/SnowballStemmer.dart

```diff
@@ -0,0 +1,574 @@
+/// The algorithm for English is documented here:
+///    Porter, M. \"An algorithm for suffix stripping.\"
+///    Program 14.3 (1980): 130-137.
+///
+/// The algorithms have been developed by Martin Porter.
+/// These stemmers are called Snowball, because Porter created
+/// a programming language with this name for creating
+/// new stemming algorithms. There is more information available
+/// at http://snowball.tartarus.org/
+///
+/// A detailed description of the English stemming algorithm can be found under
+/// http://snowball.tartarus.org/algorithms/english/stemmer.html
+
+class SnowballStemmer {
+  final _vowels = "aeiouy";
+  final _doubleConsonants = [
+    "bb",
+    "dd",
+    "ff",
+    "gg",
+    "mm",
+    "nn",
+    "pp",
+    "rr",
+    "tt"
+  ];
+  final _liEnding = "cdeghkmnrt";
+  final _step0Suffixes = ["'s'", "'s", "'"];
+  final _step1aSuffixes = ["sses", "ied", "ies", "us", "ss", "s"];
+  final _step1bSuffixes = ["eedly", "ingly", "edly", "eed", "ing", "ed"];
+  final _step2Suffixes = [
+    "ization",
+    "ational",
+    "fulness",
+    "ousness",
+    "iveness",
+    "tional",
+    "biliti",
+    "lessli",
+    "entli",
+    "ation",
+    "alism",
+    "aliti",
+    "ousli",
+    "iviti",
+    "fulli",
+    "enci",
+    "anci",
+    "abli",
+    "izer",
+    "ator",
+    "alli",
+    "bli",
+    "ogi",
+    "li",
+  ];
+  final _step3Suffixes = [
+    "ational",
+    "tional",
+    "alize",
+    "icate",
+    "iciti",
+    "ative",
+    "ical",
+    "ness",
+    "ful",
+  ];
+  final _step4Suffixes = [
+    "ement",
+    "ance",
+    "ence",
+    "able",
+    "ible",
+    "ment",
+    "ant",
+    "ent",
+    "ism",
+    "ate",
+    "iti",
+    "ous",
+    "ive",
+    "ize",
+    "ion",
+    "al",
+    "er",
+    "ic",
+  ];
+
+  final Map<String, String> _specialWords = {
+    "skis": "ski",
+    "skies": "sky",
+    "dying": "die",
+    "lying": "lie",
+    "tying": "tie",
+    "idly": "idl",
+    "gently": "gentl",
+    "ugly": "ugli",
+    "early": "earli",
+    "only": "onli",
+    "singly": "singl",
+    "sky": "sky",
+    "news": "news",
+    "howe": "howe",
+    "atlas": "atlas",
+    "cosmos": "cosmos",
+    "bias": "bias",
+    "andes": "andes",
+    "inning": "inning",
+    "innings": "inning",
+    "outing": "outing",
+    "outings": "outing",
+    "canning": "canning",
+    "cannings": "canning",
+    "herring": "herring",
+    "herrings": "herring",
+    "earring": "earring",
+    "earrings": "earring",
+    "proceed": "proceed",
+    "proceeds": "proceed",
+    "proceeded": "proceed",
+    "proceeding": "proceed",
+    "exceed": "exceed",
+    "exceeds": "exceed",
+    "exceeded": "exceed",
+    "exceeding": "exceed",
+    "succeed": "succeed",
+    "succeeds": "succeed",
+    "succeeded": "succeed",
+    "succeeding": "succeed",
+  };
+
+  String _r1;
+  String _r2;
+
+  String _word;
+
+  String stem(String origWord) {
+    _word = origWord.toLowerCase();
+
+    // TODO(jeffbailey): Check stopwords
```

Comment:
Still a to do?

Reply:
Yes.  I think that there should probably be a central stopwords set in the package rather than doing this per-stemmer.  I have an external set in my package that I'm using for now.